From 43a6f36973d290bacd967c65876edc99c17bcda3 Mon Sep 17 00:00:00 2001
From: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Date: Mon, 29 Jul 2024 17:07:12 +0200
Subject: [PATCH 1101/1695] clk: stm32mp15: enable all outputs (PQR) at the PLL
 configuration

During the initialization of the clock tree, if a PLL is enabled, also
enable the PQR outputs.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git eddcd2cf9788a21e12356b044402af3da92b4078 clk: stm32mp15: enable all outputs (PQR) at the PLL configuration]
Signed-off-by: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Change-Id: Ibb1c8c0ea4e06c6906c67928cac7614408ecd208
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/395686
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/drivers/clk/clk-stm32mp15.c | 30 +++---------------------------
 1 file changed, 3 insertions(+), 27 deletions(-)

diff --git a/core/drivers/clk/clk-stm32mp15.c b/core/drivers/clk/clk-stm32mp15.c
index 6ad9eb6bb..cf4977c68 100644
--- a/core/drivers/clk/clk-stm32mp15.c
+++ b/core/drivers/clk/clk-stm32mp15.c
@@ -3409,28 +3409,6 @@ static int clk_stm32_pll_init_switch_to_hsi_clk_system(int mux_sys)
 	return sel;
 }
 
-static uint32_t
-clk_stm32_pll_backup_output_diven(const struct stm32_clk_pll *pll)
-{
-	struct clk_stm32_priv *priv = clk_stm32_get_priv();
-	uintptr_t addr = priv->base + pll->reg_pllxcr;
-
-	return io_read32(addr + RCC_OFFSET_PLLXCR) &
-			 (RCC_PLLNCR_DIVPEN | RCC_PLLNCR_DIVQEN |
-			  RCC_PLLNCR_DIVREN);
-}
-
-static void clk_stm32_pll_restore_output_diven(const struct stm32_clk_pll *pll,
-					       uint32_t value)
-{
-	struct clk_stm32_priv *priv = clk_stm32_get_priv();
-	uintptr_t addr = priv->base + pll->reg_pllxcr;
-	const uint32_t mask = RCC_PLLNCR_DIVPEN | RCC_PLLNCR_DIVQEN |
-			      RCC_PLLNCR_DIVREN;
-
-	io_clrsetbits32(addr, mask, value & mask);
-}
-
 static int clk_stm32_pll_init(struct clk_stm32_priv *priv, int pll_idx,
 			      struct stm32_pll_dt_cfg *pll_conf)
 {
@@ -3438,7 +3416,6 @@ static int clk_stm32_pll_init(struct clk_stm32_priv *priv, int pll_idx,
 	int config_on_the_fly = -1;
 	int ret = 0;
 	uint8_t sel = 0;
-	uint32_t save_div_pqr_en = 0;
 	int mux_system[] = { MUX_MPU, MUX_AXI, MUX_MCU, -1 };
 	int mux_sys  = mux_system[pll_idx];
 
@@ -3447,9 +3424,6 @@ static int clk_stm32_pll_init(struct clk_stm32_priv *priv, int pll_idx,
 	if (ret != 0)
 		return ret;
 
-	/* Backup status of DIV DIVPEN / DIVQEN / DIVREN */
-	save_div_pqr_en = clk_stm32_pll_backup_output_diven(pll);
-
 	if (config_on_the_fly == -1) {
 		/* Make a backup to the current parent and switch to HSI */
 		sel = clk_stm32_pll_init_switch_to_hsi_clk_system(mux_sys);
@@ -3481,7 +3455,9 @@ static int clk_stm32_pll_init(struct clk_stm32_priv *priv, int pll_idx,
 		if (stm32_gate_rdy_enable(pll->gate_id))
 			return -1;
 
-		clk_stm32_pll_restore_output_diven(pll, save_div_pqr_en);
+		io_setbits32(priv->base + pll->reg_pllxcr, RCC_PLLNCR_DIVPEN |
+						   RCC_PLLNCR_DIVQEN |
+						   RCC_PLLNCR_DIVREN);
 	}
 
 	if ((config_on_the_fly == -1) && (mux_sys != -1)) {
-- 
2.47.2

